File: app/javascript/src/components/Channels/SessionApproval/session_approval_channel.js
import {
  handleWaitingParticipantJoined,
  handleWaitingParticipantLeft,
} from "components/Jitsi";
import consumer from "../consumer";

export const practitionerSubscription = ({ roomName, jitsiDispatch }) => {
  const subscription = consumer.subscriptions.create(
    {
      channel: "SessionApprovalChannel",
      room_id: roomName,
      role: "practitioner",
    },
    {
      connected() {
        this.practitionerIsOnline();
      },
      disconnected() {
        logger.log("practitioner is disconnected");
      },
      received(data) {
        if (!data.error) {
          if (data.participant_room) {
            handleWaitingParticipantJoined(data, jitsiDispatch);
          }
        } else {
          logger.log(data, "data from error");
        }
      },
      admit_participant(participant_room) {
        this.perform("admit_participant", {
          permission_granted: true,
          participant_room: participant_room,
        });
        handleWaitingParticipantLeft({ participant_room }, jitsiDispatch);
      },
      practitionerLeft() {
        this.perform("practitioner_left", {
          practitioner_left: true,
          practitioner_room_name: roomName,
        });
      },
      practitionerIsOnline() {
        this.perform("practitioner_is_online", {
          practitioner_room_name: roomName,
        });
      },
      practitionerStartedSession() {
        this.perform("practitioner_started_session", {
          practitioner_room_name: roomName,
        });
      },
    }
  );

  return subscription;
};

export const participantSubscription = ({
  roomId,
  participantDispatch,
  practitionerRoomName,
}) => {
  const subscription = consumer.subscriptions.create(
    {
      channel: "SessionApprovalChannel",
      room_id: roomId,
      role: "participant",
    },
    {
      connected() {
        this.getSessionState();
      },
      received(data) {
        if (!data.error) {
          if (data.permission_granted) {
            participantDispatch({
              type: "SET_PERMISSION",
              payload: { permissionGranted: data.permission_granted },
            });
          }
          if (data.practitioner_join_status) {
            participantDispatch({
              type: "PRACTITIONER_IS_ONLINE",
              payload: {
                isPractitionerOnline: data.is_practitioner_online,
                practitionerStartedSession: data.is_session_started,
              },
            });
          }
        } else {
          logger.log(data, "participant received data from error");
        }
      },
      getSessionState() {
        this.perform("practitioner_join_status", {
          practitioner_room_name: practitionerRoomName,
        });
      },
      disconnected() {
        logger.log("participant is disconnected");
      },
    }
  );

  return subscription;
};

File: app/channels/session_approval_channel.rb
# frozen_string_literal: true

class SessionApprovalChannel < ApplicationCable::Channel
  def subscribed
    stream_from "session_approval_channel_#{params[:room_id]}"
    load_practitioner if params[:role] == "practitioner"
  end

  def admit_participant(data)
    ActionCable.server.broadcast "session_approval_channel_#{data["participant_room"]}", { permission_granted: data["permission_granted"] }
  end

  def practitioner_left(data)
    @practitioner.toggle!(:is_session_live) if @practitioner.is_session_live?
    ActionCable.server.broadcast "conference_state_channel_#{data["practitioner_room_name"]}", { practitioner_left: data["practitioner_left"], practitioner_room_name: data["practitioner_room_name"] }
  end

  def practitioner_is_online(data)
    @practitioner.toggle!(:is_active) unless @practitioner.is_active?
    ActionCable.server.broadcast "conference_state_channel_#{data["practitioner_room_name"]}", { is_practitioner_online: true, practitioner_room_name: data["practitioner_room_name"] }
  end

  def practitioner_join_status(data)
    practitioner = User.find_by(room_name: data["practitioner_room_name"])
    ActionCable.server.broadcast "session_approval_channel_#{params[:room_id]}", { is_practitioner_online: practitioner.is_active?, practitioner_join_status: true, is_session_started: practitioner.is_session_live? }
  end

  def practitioner_started_session(data)
    @practitioner.toggle!(:is_session_live) unless @practitioner.is_session_live?
    ActionCable.server.broadcast "conference_state_channel_#{data["practitioner_room_name"]}", { practitioner_room_name: data["practitioner_room_name"], is_session_started: true }
  end

  def unsubscribed
    if params[:role] == "practitioner"
      @practitioner.toggle!(:is_session_live) if @practitioner.is_session_live?
      @practitioner.toggle!(:is_active) if @practitioner.is_active?
      ActionCable.server.broadcast "conference_state_channel_#{params[:room_id]}", { practitioner_left: true, practitioner_room_name: params[:room_id] }
    end
  end

  private

    def load_practitioner
      @practitioner = User.find_by(room_name: params[:room_id])
    end
end
